File: api/download-mp3.py
# File: api/download-mp3.py
from flask import Flask, request, jsonify
from yt_dlp import YoutubeDL

app = Flask(__name__)

@app.route('/api/download-mp3', methods=['GET'])
def download_mp3():
    video_id = request.args.get('videoId')
    if not video_id:
        return jsonify({"success": False, "error": "Missing video ID."}), 400

    url = f"https://www.youtube.com/watch?v={video_id}"

    try:
        # yt-dlp options to get the best audio-only format
        ydl_opts = {
            'format': 'bestaudio[ext=m4a]',
            'quiet': True,
        }

        with YoutubeDL(ydl_opts) as ydl:
            info_dict = ydl.extract_info(url, download=False)
            # The direct download URL is in the 'url' key of the extracted info
            download_url = info_dict.get("url")

            if not download_url:
                raise Exception("Could not find a suitable download link.")

            return jsonify({"success": True, "downloadUrl": download_url})

    except Exception as e:
        print(f"--- DETAILED DOWNLOAD ERROR --- \n{e}\n--- END ERROR ---")
        return jsonify({"success": False, "error": f"Failed to get download link: {str(e)}"}), 500

# This is the entry point for Vercel
def handler(request, context):
    return app(request, context)

File: api/download-mp3.js
// File: /api/download-mp3.js (Final Corrected Version)
const play = require('play-dl');
const ffmpeg = require('fluent-ffmpeg');
const ffmpegInstaller = require('@ffmpeg-installer/ffmpeg');

// Point fluent-ffmpeg to the bundled binary
ffmpeg.setFfmpegPath(ffmpegInstaller.path);

// Function to apply user cookies to play-dl
const setAuth = async () => {
    const cookie = process.env.YOUTUBE_COOKIES;
    if (cookie && typeof cookie === 'string') {
        try {
            const sanitizedCookie = cookie.replace(/[^\x20-\x7E]/g, '');
            await play.setToken({
                youtube: {
                    cookie: sanitizedCookie,
                },
            });
            console.log('Successfully attempted to set YouTube authentication token.');
        } catch (e) {
            console.error('Error setting YouTube token:', e.message);
        }
    } else {
        console.warn('Warning: YOUTUBE_COOKIES environment variable not found or is not a string.');
    }
};

module.exports = async (req, res) => {
    // Standard CORS headers
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    if (req.method === 'OPTIONS') {
        return res.status(200).end();
    }

    try {
        await setAuth();

        const { videoId, title } = req.query;

        // Rigorous check for videoId
        if (!videoId || typeof videoId !== 'string' || videoId === 'undefined') {
            console.error(`Invalid videoId received: ${videoId}`);
            return res.status(400).json({ success: false, error: 'Invalid or missing video ID.' });
        }

        const cleanTitle = (title || 'audio').replace(/[<>:"/\\|?*]+/g, '_');
        const videoUrl = `https://www.youtube.com/watch?v=${videoId}`;

        console.log(`Attempting to stream directly from URL: ${videoUrl}`);
        
        // Reverting to the more direct stream method.
        const stream = await play.stream(videoUrl, {
            discordPlayerCompatibility: true
        });

        res.setHeader('Content-Disposition', `attachment; filename="${cleanTitle}.mp3"`);
        res.setHeader('Content-Type', 'audio/mpeg');

        ffmpeg(stream.stream)
            .audioBitrate(128)
            .toFormat('mp3')
            .on('error', (err) => {
                console.error('FFmpeg error:', err.message);
                if (!res.headersSent) {
                    res.status(500).send(`FFmpeg error: ${err.message}`);
                }
            })
            .pipe(res, { end: true });

    } catch (error) {
        console.error('--- DETAILED DOWNLOAD ERROR ---');
        console.error(error);
        console.error('--- END DETAILED ERROR ---');
        
        if (!res.headersSent) {
            res.status(500).json({ 
                success: false, 
                error: `Failed to process download. The video may be unavailable or blocked. Please check the server logs for details.` 
            });
        }
    }
};